### PR TITLE
JENKINS-51063 the PushHookProcessor does not handle correctly webhook of kind tag

### PR DESCRIPTION
**Pull request by nfalco79**
Manage TagSCMHead correctly when bitbucket payload webhook is of king create tag.
Fix parsing of ISO8601 date in BB payload. Handle also non ISO8601 standard date for updateOn field.
To avoid too regression only the new date fields are of kind Date the other one are keep as is. If changed to Date in the future they works, otherwise in case the Bitbucket date format can be enhanced to handle other cases.

### CODE REVIEW

**Comment by ryanewtaylor** [comment]
When might this PR be merged and a new version of the plugin released? I'm unable to find release cycle details elsewhere.

**Comment by nfalco79** [comment]
> When might this PR be merged and a new version of the plugin released? I'm unable to find release cycle details elsewhere.

That I know the release of each plugin is at the discretion of the maintainer

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java [hunk 1]
```diff
@@ -62,10 +95,8 @@ public static String toJson(Object value) throws IOException {
...
         ObjectMapper mapper = new ObjectMapper();
-        SimpleDateFormat format = new SimpleDateFormat(DATE_FORMAT);
-        format.setTimeZone(TimeZone.getTimeZone("UTC"));
-        mapper.setDateFormat(format);
+        mapper.setDateFormat(new BitbucketDateFormat());
         mapper.configure(DeserializationConfig.Feature.FAIL_ON_UNKNOWN_PROPERTIES, false);
         return mapper;
     }
-}
+}
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Why did it deserve a final new line deletion 😇

File: src/test/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketDateFormatTest.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    public void test_custom_bitbucket_ISO8601_date_format() {
+        ParsePosition pos = new ParsePosition(0);
+        BitbucketDateFormat bdf = new BitbucketDateFormat();
+
+        assertThat(bdf.parse("2018-09-12T14:56:34.008922+00:00", pos), is(DateUtils.getDate(2018, 9, 12, 14, 56, 34, 8)));
+        assertThat(bdf.parse("2018-09-12T14:56:34.008922-00:00", pos), is(DateUtils.getDate(2018, 9, 12, 14, 56, 34, 8)));
+        assertThat(bdf.parse("2018-09-12T14:56:34.008922Z", pos), is(DateUtils.getDate(2018, 9, 12, 14, 56, 34, 8)));
+    }
+
+}
```
Comment (casz) [inline src/test/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketDateFormatTest.java]: final line police 👮

File: src/test/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Test
+    public void get_repository_parse_correctly_date_from_cloud() throws Exception {
+        BitbucketCloudRepository repository = JsonParser.toJava(loadPayload("getRepository"), BitbucketCloudRepository.class);
+        assertNotNull("update on date is null", repository.getUpdatedOn());
+        Date date = DateUtils.getDate(2018, 4, 27, 15, 32, 8, 356);
+        assertThat(repository.getUpdatedOn().getTime(), CoreMatchers.is(date.getTime()));
+    }
+
+}
```
Comment (casz) [inline src/test/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest.java]: final line police 👮

File: src/test/java/com/cloudbees/jenkins/plugins/bitbucket/client/DateUtils.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    public static Date getDate(int year, int month, int day, int hours, int minutes, int seconds, int milliseconds, String timezoneId) {
+        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone(timezoneId));
+        cal.set(year, month - 1, day, hours, minutes);
+        cal.set(Calendar.SECOND, seconds);
+        cal.set(Calendar.MILLISECOND, milliseconds);
+        return cal.getTime();
+    }
+
+}
```
Comment (casz) [inline src/test/java/com/cloudbees/jenkins/plugins/bitbucket/client/DateUtils.java]: final line police 👮

File: src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest/getRepositoryPayload.json [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+			}
+		}
+	},
+	"updated_on": "2018-04-27T15:32:08.356155+00:00",
+	"size": 136381135,
+	"type": "repository",
+	"slug": "tds.cm.maven.plugins-java",
+	"is_private": true,
+	"description": "The acme maven plugins for build system"
+}
```
Comment (casz) [inline src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest/getRepositoryPayload.json]: final line police 👮

File: src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPushEventTest/newTagPayload.json [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      },
+      "html": {
+        "href": "https://bitbucket.org/nfalco79/"
+      },
+      "avatar": {
+        "href": "https://bitbucket.org/account/nfalco79/avatar/32/"
+      }
+    }
+  }
+}
```
Comment (casz) [inline src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/events/BitbucketCloudPushEventTest/newTagPayload.json]: final line police 👮

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java [hunk 7]
```diff
@@ -23,25 +23,58 @@
...
 
+import com.google.common.base.Charsets;
+
 /**
  * Jackson based JSON parser
  */
 @Restricted(NoExternalUse.class)
 public final class JsonParser {
-    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSS";
+    /*package*/ static class BitbucketDateFormat extends ISO8601DateFormat {
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Might need to change this after updating Jackson 😇

Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Actually I think you can skip this function and just use `StdDateFormat` from Jackson which support the case your custom parser is trying to fix. The StdDateFormat allows upto 9 milliseconds anything longer it complains which I would too 😅

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java (hunk 1 above)
Comment (nfalco79) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: In my formatter I usually remove new line at the end. Is available an eclipse formatter for jenkins projects?

File: src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest/getRepositoryPayload.json (hunk 5 above)
Comment (nfalco79) [reply src/test/resources/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClientTest/getRepositoryPayload.json]: also in json test resource files?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java (hunk 7 above)
Comment (nfalco79) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/JsonParser.java]: Ok your right, with new Jackson version from master the StdDateFormat works correctly also in the above cases (ok)

**Comment by nfalco79** [comment]
Code changed as requested

**Comment by casz** [comment]
@nfalco79 I will switch [jenkinsci/jackson2-api-plugin](https://github.com/jenkinsci/jackson2-api-plugin) for now.

However I would require it to be updated to version 2.9 for #107 in the near future
Will do that in a separate PR probably at the same time I would also think it is time to release with java 8, git plugin 4.0 is currently in beta with java 8 as target.

So for now the project would be java 7 and we can release one with the features that don't depend on java 8 dependencies 😅

**Comment by nfalco79** [comment]
ok maybe your comment was related to my comment of #143 than this PR

**Comment by casz** [comment]
Test fails with jackson-api-plugin 2.7.3 which is the mimimum version for java 7

**Comment by nfalco79** [comment]
Please move discussion to the #143 or this thread will be incomprehensible for new reader

**Comment by nfalco79** [comment]
Restore DateFormatter untill the version of jackson2-api is greater to 2.9.2

**Comment by casz** [comment]
@nfalco79 would you mind rebasing to get Jenkins ci to run test?

**Comment by casz** [comment]
sadly there was a test failure 😢

**Comment by nfalco79** [comment]
> sadly there was a test failure 😢

yes my fault, due to continue refactor the test case of date format points to the StdDateFormat instead of my BitbucketDateFormat

**Comment by chris5287** [comment]
@casz any idea when a new release of this plugin might be ready?

**Comment by casz** [comment]
When #130, #138 and #145 are merged

**Comment by Barathwaja** [comment]
Updated the plugin to 2.2.13 but still it _doesn't trigger_, and the project is a multi-branch pipeline.
Is the below config in jenkins correct? and in BB Hook said as **Build All with empty values**

Jenkins - 2.136
Bitbucket - 5.12.2

image

**Comment by nfalco79** [comment]
what do you expect when push a new tag?

1. a build is triggered somewhere?
2. a job for that tag shoud appear in the Tags tab?
3. other

**Comment by Barathwaja** [comment]
When Branch Trigger the build will start right similar think should happen for Tag also, build should start on new tag / existing tag.

**Comment by nfalco79** [comment]
To archive your goal another plugin is needed. this plugin must implement a specific branch strategy to allow trigger build on tag (i do not know if exist). If you want trigger a build manually than enable discovery tag trait. That's all

**Comment by Barathwaja** [comment]
@nfalco79  I enabled **Discover tags** and also configured respective tag wildcard expression in _Bitbucket Server Hooks_ , after creating new tag/pushing commits to tag its not building or recognizing in Jenkins. Only Branches are recognized since i have given **Discover Branches** in that.

**Comment by vinnyvinny1989** [comment]
Does anyone know if a Bitbucket Branch Source Plugin can start building job after i create and push tag from git?

**Comment by Barathwaja** [comment]
Jenkins is not triggering the build if tag is created and pushed.

**Comment by vinnyvinny1989** [comment]
> 
> 
> Jenkins is not triggering the build if tag is created and pushed.

Maybe you know a way how to resolve this?)

**Comment by Barathwaja** [comment]
No idea mate..!! 😕 anyone?

**Pull request merged**